bindings/java/src/org/sleuthkit/datamodel/SleuthkitJNI.java

```diff
@@ -231,17 +235,18 @@ public void run(String deviceId, String[] imageFilePaths) throws TskCoreExceptio
 				if (0 != tskAutoDbPointer) {
 					throw new TskCoreException("Add image process already started");
 				}
-
 				long imageHandle = openImage(imageFilePaths, false);
-				
-				synchronized (this) {					
-					tskAutoDbPointer = initAddImgNat(caseDbPointer, timezoneLongToShort(timeZone), addUnallocSpace, skipFatFsOrphans);
+				synchronized (this) {
+					if (!isCanceled) {
+						tskAutoDbPointer = initAddImgNat(caseDbPointer, timezoneLongToShort(timeZone), addUnallocSpace, skipFatFsOrphans);
+					}
 				}
 				if (0 == tskAutoDbPointer) {
 					throw new TskCoreException("initAddImgNat returned a NULL TskAutoDb pointer");
 				}
-
-				runAddImgNat(tskAutoDbPointer, deviceId, imageHandle, timeZone, imageWriterPath);
+				if (!isCanceled) {
+					runAddImgNat(tskAutoDbPointer, deviceId, imageHandle, timeZone, imageWriterPath);
+				}
```

Comment:
@bcarrier, @APriestman and @wschaeferB, I am still finding this code difficult to reason about to convince myself it is thread-safe. I am wondering if by attempting to avoid synchronizing the stop() method, 
we are attempting to slice things rather fine for a class that is already mostly synchronized on "this." The optimization of leaving one method out, if it actually gains us anything (I'm not sure),
may be uneccessary given the infrequent use of this class. I propose the following for safety and clarity:  

1. In the run() method, move lines 235 to 246 into the existing synchronized block.
2. Make the stop() method synchronized, just like the revert(), commit(), and getCurrentDirectory() methods already are.


Then the tskAutoDbPointer field is fully guarded by "this" (and can be so documented with an annotation), and does not need to be volatile. 
The "don't break this" comments can then be removed from stop(). And we can reason thus: 

1. If stop() is called before run(), run() cannot proceed to set tskAutoDbPointer before the isCanceled flag is set, which means that the read on line 247 cannot happen before the write on line 262.
2. If run() is called before stop(), we are guaranteed that run() will see a non-null tskAutoDbPointer and will call stopAddImagNat.

Am I correct? Your thoughts?